core/src/main/scala/org/apache/spark/shuffle/ShuffleMergerDiscovery.scala

```diff
@@ -0,0 +1,37 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle
+
+import org.apache.spark.annotation.{Experimental, Since}
+import org.apache.spark.storage.BlockManagerId
+
+/**
+ * :: Experimental ::
+ * Pluggable interface to get merger locations for Push based shuffle.
+ */
+@Experimental
+@Since("3.1.0")
+trait ShuffleMergerDiscovery {
```

Comment:
It seems like even though we have introduced this API, we are still adding similar API to `SchedulerBackend`.
What is needed from `SchedulerBackends`? Is it just a list of hosts that should be excluded for mergeLocations?
In that case, should we change the API in the `SchedulerBackends` to just retrieve that list?

Reply:
Currently this way for other schedulers like (mesos/k8s etc) push based shuffle would be disabled and also different scheduler backends can have different implementations for fetching merger locations. Like I said above, I feel better to have more feedback from OS than doing more iterations on this API now. Min was also fine with that. Let me know if you have any concerns.

Reply:
@venkata91 I am fine with getting more feedback from OS. However, I don't see a value of adding this API. From your response it seems that `getMergerLocations` API should be part of the scheduler backends then we probably shouldn't be adding this API. 
cc. @Victsm @mridulm

Reply:
@otterc This API would serve the purpose of getting merger locations from an external implementation but different scheduler backends can choose to have their own implementation or reuse the existing implementation for getting merger locations. I feel merger locations should be abstracted away from DAGScheduler since we have multiple SchedulerBackends and having only one implementation might be too strict.

Reply:
I feel that we should also include application ID and shuffle ID in the API.
This is not needed now, but might be necessary if people set out to build an external merger location discovery service.

Reply:
Makes sense, then I will resolve this conversation for now.

Reply:
Since this API itself at the current stage is still very lightweight, I think what might be more reasonable is to not include `ShuffleMergerDiscovery` initially in the PR.
Let's send out the PR w/o this API, and then we can raise the discussion in the community about the need of it.
I think this way is probably better for the community to review and chime-in.
Currently, adding this API into the PR could cause some confusions.